src/models/right_view/vae/vae_unet.py

```diff
@@ -0,0 +1,169 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class VariationalUNet(nn.Module):
+    def __init__(
+            self,
+            input_channels: int,
+            output_channels: int = 3,
+            enc_out_dim: int = 512,
+            latent_dim: int = 256,
+            kl_coeff: float = 0.01,
+            num_layers: int = 5,
+            features_start: int = 64,
+            bilinear: bool = False
+    ):
+        super().__init__()
+        self.num_layers = num_layers
+        self.input_channels = input_channels
+        self.kl_coeff = kl_coeff
+
+        layers = [DoubleConv(self.input_channels, features_start)]
+
+        feats = features_start
+        for _ in range(num_layers - 1):
+            layers.append(Down(feats, feats * 2))
+            feats *= 2
+
+        for _ in range(num_layers - 1):
+            layers.append(Up(feats, feats // 2, bilinear))
+            feats //= 2
+
+        layers.append(nn.Conv2d(feats, output_channels, kernel_size=1))
+
+        self.layers = nn.ModuleList(layers)
+
+        self.fc = nn.Linear(1024 * 12 * 24, enc_out_dim)
+
+        self.fc_mu = nn.Linear(enc_out_dim, latent_dim)
+        self.fc_logvar = nn.Linear(enc_out_dim, latent_dim)
+
+        # initialize weights to try prevent large std and nan values
+        self.fc_logvar.weight.data.uniform_(-0.001, 0.001)
+
+        self.projection_1 = nn.Linear(latent_dim, 1024 * 12 * 24)
+        self.projection_2 = nn.Sequential(
+            nn.Conv2d(2 * 1024, 1024, kernel_size=3, padding=1),
+            nn.BatchNorm2d(1024),
+            nn.ReLU(inplace=True)
+        )
+
+    def gaussian_like(self, mean, logscale, sample):
+        scale = torch.exp(logscale)
+        dist = torch.distributions.Normal(mean, scale)
+        log_pxz = dist.log_prob(sample)
+        log_pxz = log_pxz.sum(dim=(1,2,3))
+        return log_pxz
+
+    def forward(self, x):
+        # down path / encoder
+        xi = [self.layers[0](x)]
+        for layer in self.layers[1:self.num_layers]:
+            output = layer(xi[-1])
+            xi.append(output)
+
+        # embedding
+        emb = xi[-1]
+        emb = emb.view(emb.size(0), -1)
+        emb = self.fc(emb)
+
+        # variational
+        mu = self.fc_mu(emb)
+        logvar = self.fc_logvar(emb)
+
+        # kl
```

Comment:
I think we should split this into another function so we can change functionality during eval. Something like:
```
  def reparameterize(self, mu, logvar):
    if self.training:
      std = torch.exp(0.5*logvar)
      eps = torch.randn_like(std)
      z = mu + std*eps
    else:
      z = mu
    return z
```

Reply:
added this too